### PR TITLE
exec: initial commit of execgen tool

### PR DESCRIPTION
**Pull request by solongordon**
Execgen will be our tool for generating templated code necessary for
columnarized execution. So far it only generates the
EncDatumRowsToColVec function, which is used by the columnarizer to
convert a RowSource into a columnarized Operator.

Release note: None

### CODE REVIEW

**Review by jordanlewis — CHANGES_REQUESTED** [review]
Great!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-:-LP3N5Lj5Uriq6MSq9LG:btojbfd)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/.gitignore, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP3Mjdd2E2TMtTNhXvy:-LP3Mjdd2E2TMtTNhXvz:bqbejfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c55c373d30696a05b205af71e5d79ca29e32bf5/pkg/sql/exec/.gitignore#L1)):*
> ```
> rowstovec.og.go
> ```

I'd just make this *.og.go so we don't have to keep editing it
___
*[pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go, line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP3H3I7-yyIa5r6MAGL:-LP3H3I7-yyIa5r6MAGM:b-7anjct) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c55c373d30696a05b205af71e5d79ca29e32bf5/pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go#L34)):*
> ```Go
> 
> // EncDatumRowsToColVec converts one column from EncDatumRows to a column
> // vector.
> ```

Can you add a comment explaining the columnIdx parameter? Seems like something that could end up being confusing for newcomers one day (all those index-to-index maps, for example).
___
*[pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP3IgryE5o_0PW4wqew:-LP3IgryE5o_0PW4wqex:b-5fu5qh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c55c373d30696a05b205af71e5d79ca29e32bf5/pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go#L44)):*
> ```Go
> 	nRows := uint16(len(rows))
> 	{{range .}}
> 	if columnType.SemanticType == sqlbase.{{.SemanticType}} && columnType.Width == {{.Width}} {
> ```

I'm a little suspicious about this, because I don't think Width has semantic meaning for most types - just INT, DECIMAL, CHAR, and BINARY according to the comment in structured.proto.

I suppose if they're just left zeroed on both sides (in the columnConversion as well as the ColumnType) this will just be a no-op... but since we're in template land, I feel like we might as well just remove the width equality methods since we do have the chance to do so.


Finally, I think the code will be more efficient with a `switch` on `SemanticType` the way it's currently implemented in `columnarizer.go`. That's because `switch` does a binary search to find the correct case statement. We could even use a lookup table directly, since ColumnType_SemanticType (though an int32 for some reason) doesn't have many values. Then again, this is just once per batch, so maybe who cares... I'd definitely be motivated to fix this myself, but I guess we've got more to do for a prototype than to shave every last batch-level microsecond.

tl;dr this whole comment is optional for now but would definitely yield less work for the CPU to do
___
*[pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP3IGE7E_UTw9KkCLv4:-LP3IGE7E_UTw9KkCLv5:b-e42dqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c55c373d30696a05b205af71e5d79ca29e32bf5/pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go#L56)):*
> ```Go
> 				vec.SetNull(i)
> 			} else {
> 				{{if .HasSetMethod}}
> ```

Nice - this is a lot less gross than what I was fearing, now that I see it written down. I suppose a similar approach could probably go well for the infix operators too - just a simple if.
___
*[pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP3KPh_-Kmdt-6FZGrw:-LP3KPh_-Kmdt-6FZGrx:bimcis7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c55c373d30696a05b205af71e5d79ca29e32bf5/pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go#L88)):*
> ```Go
> }
> 
> var columnConversions = []columnConversion{
> ```

I would consider making this slice generated from a map in an `init` function.

The key of the map can be the actual semantic type value - e.g. `ColumnType_BOOL`. Protobuf provides an autogenerated map of value to name (`ColumnType_SemanticType_name`) that gives you back `"BOOL"` for `ColumnType_Bool`. For the `ExecType`,  I would consider using `FromColumnType` on the key of the map to get the actual value of the execType, and then adding a new map to `types.go` similar to `ColumnType_SemanticType_name` that gives you back the actual variable name.

For `HasSetMethod`, you can use the reflect package. Do `newMemColumn` on the type, then `reflect.TypeOf` the inner column of the type, then ask it if it `reflect.Implements` a marker interface that you can give to all of those vector types (`type HasSet interface { HasSetMarker() }`). 

For `DatumToPhysicalFn` I'm sure there's something you can do too.

Perhaps the last two are too high effort and tricky to get right, but I don't think there's a strong reason not to do it for the SemanticType and ExecType names, and I think we'll likely reuse that work in other templates.

**Comment by jordanlewis** [comment]
I have a few thoughts that I'm jotting down here, don't take these comments as work to be done on the current PR though.

Instead of hardcoding the files that will be generated in main.go, like optgen does, we probably want to have that be autogenerated as well, since we'll have a long list of generated code.

I'm envisioning a `registerGenerator` function that each operator will call in its `init` function that adds a `generator` to a list of things that `execgen` can generate. That will minimize the number of things that has to change each time a new operator template is added - and I think there will be lots of these templates over time.

Also, there are some common mappings that we'll need over and over again in templates, like the name of the vector accessor for each type (currently always exactly equal to the type enum name, I guess), the semantic type, and the Go type associated with the exec type (e.g. `int32` for `exec.Int32`). I'm hoping the `columnConversions` slice or something like it can grow over time into a generic all-types slice that every operator can iterate over in the same way to generate all of its variants. Not sure if it'll work out in practice, though.

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31554)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/.gitignore, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP3Mjdd2E2TMtTNhXvy:-LP6XkdP8BRg5pOdKXhr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c55c373d30696a05b205af71e5d79ca29e32bf5/pkg/sql/exec/.gitignore#L1)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'd just make this *.og.go so we don't have to keep editing it
</blockquote></details>

Done.
___
*[pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go, line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP3H3I7-yyIa5r6MAGL:-LP6XzIL3m8rPWzTmu0E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c55c373d30696a05b205af71e5d79ca29e32bf5/pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go#L34)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Can you add a comment explaining the columnIdx parameter? Seems like something that could end up being confusing for newcomers one day (all those index-to-index maps, for example).
</blockquote></details>

Done.
___
*[pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP3IgryE5o_0PW4wqew:-LP6Y7zS3im3K3iJC_io:b-4jiicf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c55c373d30696a05b205af71e5d79ca29e32bf5/pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go#L44)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'm a little suspicious about this, because I don't think Width has semantic meaning for most types - just INT, DECIMAL, CHAR, and BINARY according to the comment in structured.proto.

I suppose if they're just left zeroed on both sides (in the columnConversion as well as the ColumnType) this will just be a no-op... but since we're in template land, I feel like we might as well just remove the width equality methods since we do have the chance to do so.


Finally, I think the code will be more efficient with a `switch` on `SemanticType` the way it's currently implemented in `columnarizer.go`. That's because `switch` does a binary search to find the correct case statement. We could even use a lookup table directly, since ColumnType_SemanticType (though an int32 for some reason) doesn't have many values. Then again, this is just once per batch, so maybe who cares... I'd definitely be motivated to fix this myself, but I guess we've got more to do for a prototype than to shave every last batch-level microsecond.

tl;dr this whole comment is optional for now but would definitely yield less work for the CPU to do
</blockquote></details>

Yes, I meant to leave a TODO here. I'm doing the simple thing for now and punting on making it more efficient until we handle more types. I'll add a comment.
___
*[pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP3KPh_-Kmdt-6FZGrw:-LP6d0vD0Lb_tPos5owQ:b-pfny2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c55c373d30696a05b205af71e5d79ca29e32bf5/pkg/sql/exec/execgen/cmd/execgen/rowstovec_gen.go#L88)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I would consider making this slice generated from a map in an `init` function.

The key of the map can be the actual semantic type value - e.g. `ColumnType_BOOL`. Protobuf provides an autogenerated map of value to name (`ColumnType_SemanticType_name`) that gives you back `"BOOL"` for `ColumnType_Bool`. For the `ExecType`,  I would consider using `FromColumnType` on the key of the map to get the actual value of the execType, and then adding a new map to `types.go` similar to `ColumnType_SemanticType_name` that gives you back the actual variable name.

For `HasSetMethod`, you can use the reflect package. Do `newMemColumn` on the type, then `reflect.TypeOf` the inner column of the type, then ask it if it `reflect.Implements` a marker interface that you can give to all of those vector types (`type HasSet interface { HasSetMarker() }`). 

For `DatumToPhysicalFn` I'm sure there's something you can do too.

Perhaps the last two are too high effort and tricky to get right, but I don't think there's a strong reason not to do it for the SemanticType and ExecType names, and I think we'll likely reuse that work in other templates.
</blockquote></details>

Done. I'm saving the last two for late. (I fiddled with suing reflection for `HasSetMethod` but it wasn't feeling worth the time at this point. And I think it would require exporting all the `MemColumn` stuff.)

Also I put this logic in the generator function rather than an `init()` for now since no one else needs the `columnConversions` so far.

**Review by benesch — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31554)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[Makefile, line 1315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-LP7gKIa1tZAXPrL_XIc:-LP7gKIb-Wt00-qiRFzj:b-4tj6f2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d2038d4046c8637c53354cc2504cb8608839adc/Makefile#L1315)):*
> ```Makefile
> pkg/sql/exec/rowstovec.og.go: bin/execgen
> 	execgen -out $@ rowstovec
> ```

nit: if you go through CI again, move this above the optgen variable declarations

**Review by solongordon — COMMENTED** [review]
I went ahead and added the `registerGenerator` functionality. Also changed it to pick the generator based on the specified filename, which reduces how much we have to touch in both `execgen/main.go` and `Makefile` when we add new generators.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31554#-:-LP7vn-76-LgQuldqrbb:bc7vlvs)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by jordanlewis — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31554)* status: :shipit: complete! 0 of 0 LGTMs obtained
